node/src/Pos/Block/Network/Retrieval.hs

```diff
@@ -305,17 +306,20 @@ handleCHsValid enqueue endedRecoveryVar nodeId lcaChild newestHash = do
                 -- difficulty than ncrecoveryheader, we're
                 -- gracefully exiting recovery mode.
                 let isMoreDifficultThan b x = b ^. difficultyL >= x ^. difficultyL
-                endedRecovery <- atomically $ do
+                atomically $ do
                     mRecHeader <- tryReadTMVar recHeaderVar
                     case mRecHeader of
                         Nothing -> return False
                         Just (_, rHeader) ->
                             if any (`isMoreDifficultThan` rHeader) blocks
                                 then isJust <$> tryTakeTMVar recHeaderVar
                                 else return False
-                void $ tryPutMVar endedRecoveryVar endedRecovery
-    void $ waitForConversations its
+    results <- waitForConversations $ fmap try convs
+    let Any endedRecovery = fold $ fmap mkAny results
+    return endedRecovery
   where
+    mkAny :: Either SomeException Bool -> Any
+    mkAny = either (const (Any False)) Any
```

Comment:
Why use `try` to later discard the exception like this? Wouldn't it be easier simpler to do something like `handle (\_ -> return False`) directly?

Furthermore, discarding exceptions like this is bad, what if it's an async exception? You now can't even kill the thread. Better catch a specific set of exceptions here.